lib/image_optim/cmd.rb

```diff
@@ -15,6 +18,45 @@ def run(*args)
         success
       end
 
+      def support_timeout?
+        if defined?(JRUBY_VERSION)
+          JRUBY_VERSION >= '9.0.0.0'
+        else
+          RUBY_VERSION >= '1.9'
+        end
+      end
+
+      # Run commands using `Process.spawn`
+      # Return success status
+      # Will raise `ImageOptim::Cmd::Timeout` when command timeouts
+      def run_with_timeout(timeout, *args)
+        return run(*args) unless timeout > 0 && support_timeout?
+
+        success = false
+        init_options!(args)
+
+        begin
+          stdin, stdout, thread = Open3.popen2(*args)
+          now = Time.now
+          pid = thread[:pid]
+
+          sleep 0.001 while (Time.now - now) < timeout && thread.alive?
+
+          if thread.alive?
+            cleanup_process(pid)
+            thread.kill
+            fail Cmd::Timeout
+          else
+            success = thread.value.exitstatus.zero?
+          end
+        ensure
+          stdin.close
```

Comment:
Thank you for explaining @mpalmer :+1: 

I've gone ahead and just close `stdout` since the only method that uses this library [redirects `:out` and `:err`](https://github.com/toy/image_optim/blob/dc0d44e49684b3bd1e4f290729387c7ad18d1da0/lib/image_optim/worker.rb#L140-L152) to [Path::NULL](https://github.com/toy/image_optim/blob/dc0d44e49684b3bd1e4f290729387c7ad18d1da0/lib/image_optim/path.rb#L7-L11)